doc/rst/technotes/optimization.rst

```diff
@@ -0,0 +1,662 @@
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Issue: https://github.com/chapel-lang/chapel/issues/10160
+
+When a class contains a field that is a distributed array, a distributed
+``forall`` loop using a class instance will generate communication in
+order to read the field's value (repeatedly) even though the field's
+value should be a privatized distributed array.
+
+This issue can be avoided by creating a ``ref`` or ``const ref`` that
+refers to the distributed array. This ``ref`` or ``const ref`` can be
+created outside of the ``forall`` loop and reused within it to avoid the
+problem.
+
+Creating Too Many Distributed Objects / Unoptimized Slice Assignments
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+In principle, distributed array creation does some work on each locale. As a
+result, it's not going to go faster when adding more locales. That can
+cause performance or scaling issues if a program tries to create too many
+distributed arrays.
+
+One aspect of the current implementation is that array slices of
+distributed arrays are also distributed objects and work is required on
+each locale to create one. That means that creating a slice, as in
+``MyDistributedArray[1..10]`` can actually be quite slow.
+
+One common scenario is that data needs to be copied between regions of
+arrays. For example
+``MyDistributedArray[1..10] = MyOtherArray[11..20];``.
+The compiler can optimize this kind of
+assignment in many cases today with Array View Elision (AVE). However, in
+order to optimize it in this way, both sides of the ``=`` need to be
+slice expressions. Note that in some cases, one might create a local
+array to store the contents of a remote slice; that can be done with AVE
+by redundantly slicing the local array, e.g.:
+``var Loc:[1..10] int; Loc[1..10] = MyDistributedArray[1..10];```
+
+Unoptimized Distributions
+~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The Chapel programming language is designed to support many distributions
+for domains and arrays. That includes distributions created by users.
+However, the present situation is that performance of Chapel programs
+depends on optimizations in the implementations of these domain/array
+distributions. These optimizations are present and reasonably well tuned
+for the Block distribution. Other distributions might not be optimized
+and have scaling issues.
+
+Performance Problems with Multidimensional Zippered Iteration
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Issue: https://github.com/chapel-lang/chapel/issues/13147)
+
+Zippered iteration for multidimensional arrays/domains is much slower
+than zippered iteration for 1D arrays/domains. Since promoted calls, such
+as ``MyArray + MyOtherArray`` are implemented with zippered iteration,
+this problem also applies to that case.
+
+Potential ways to avoid this problem:
+ * express the computation with nested loops per dimension
```

Comment:
does this mean I can't iterate over the domain, and have to explicitly have N nested loops?

I for one would benefit from this information.

Reply:
You can iterate over the multidimensional domain, just can't do *zippered* iteration over the multidimensional domain.